pkg/util/fsm/fsm.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2017 ScyllaDB
+
+package fsm
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// ErrEventRejected is the error returned when the state machine cannot process
+// an event in the state that it is in.
+var ErrEventRejected = errors.New("event rejected")
+
+const (
+	// Default represents the default state of the system.
+	Default StateType = ""
+
+	// NoOp represents a no-op event.
+	NoOp EventType = "NoOp"
+)
+
+// StateType represents an extensible state type in the state machine.
+type StateType string
```

Comment:
I'd suggest rename to State, the current state looks like it's implementation specific and should not be exported.